### PR TITLE
[SPARK-241909] Border Color and Icon changes based on UX (Windows)

### PR DESCRIPTION
**Pull request by dgharde**
## Related Issue
Please use one of the well-known [github fixes keywords](https://help.github.com/en/articles/closing-issues-using-keywords) to reference
the issues fixed with this PR (eg Fixes #<github issue number>). If an issue doesn't yet exist please create one if reasonable to aid 
in issue tracking.

**NOTE**: For multiple issues resolved by this PR use the corresponding keywords **every time** in a comma-delimited list per the reference
page above.

## Description
- Changed border colors for all input elements and for the card

- Change choiceSet colors

- Fix TextArea placeholder padding

- Change Icon color for Input Number and Action ShowCard

- Remove Bottom Margin when showCard is opened

https://www.figma.com/file/fi0VC0FEmETIHTTWANqMqJ/Native-B-and-Cs?node-id=0%3A1

## Sample Card
Input Number Light
image
Input Number Dark
image

Input ChoiceSet Light
image

Input ChoiceSet Dark
image

Action ShowCard Light
image

Action ShowCard Dark
image

## How Verified
How you verified the fix, including one or all of the following: 
1. New unit tests that were added if any. If none were added please add a quick line explaining why not.
2. Existing relevant unit/regression tests that you ran
3. Manual scenario verification if any; ***Do include .gif's or screenshots of the testing you performed here if you think that it 
will aid in code reviews or corresponding fixes on other platforms for eg.***

### CODE REVIEW

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp [hunk 1]
```diff
@@ -118,10 +120,20 @@ namespace RendererQml
...
+		}
+
+		CheckShowCardInActions(card->GetActions(), context);
+		AddContainerElements(bodyLayout, card->GetBody(), context);
+		AddActions(bodyLayout, card->GetActions(), context);
         addSelectAction(uiCard, uiCard->GetId(), card->GetSelectAction(), context, hasBackgroundImage);
 
+		auto showCardsList = context->getShowCardsLoaderIdsList();
+		auto removeBottomMarginValue = RemoveBottomMarginValue(showCardsList);
+		rectangle->Property("property bool removeBottomMargin", removeBottomMarginValue);
```
Comment (dgharde) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:135]: Custom Property created to control the bottom Margin of the card.

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.h [hunk 2]
```diff
@@ -114,5 +114,12 @@ namespace RendererQml
         static std::shared_ptr<QmlTag> GetIconTag(std::shared_ptr<AdaptiveRenderContext> context);
 
         static std::shared_ptr<QmlTag> GetTextBlockMouseArea();
+        static std::shared_ptr<QmlTag> getDummyElementforNumberInput(bool isTop);
+
+        static void CheckLastBodyElementIsShowCard(std::shared_ptr<AdaptiveCards::BaseCardElement> cardElement, std::shared_ptr<AdaptiveRenderContext> context);
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.h:119]:
```suggestion
        static void ValidateLastBodyElementIsShowCard(std::shared_ptr<AdaptiveCards::BaseCardElement> cardElement, std::shared_ptr<AdaptiveRenderContext> context);
```

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp [hunk 3]
```diff
@@ -118,10 +119,20 @@ namespace RendererQml
         rectangle->Property("Layout.preferredHeight", "bodyLayout.height");
         rectangle->AddChild(bodyLayout);
 
-        AddContainerElements(bodyLayout, card->GetBody(), context);
-        AddActions(bodyLayout, card->GetActions(), context);
+		if (!card->GetBody().empty())
+		{
+			CheckLastBodyElementIsShowCard(card->GetBody().back(), context);
+		}
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:125]: You should do the if check inside `CheckLastBodyElementIsShowCard`. Would be cleaner here. Just pass `card->GetBody()` and let eh function do the validations.

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp [hunk 4]
```diff
@@ -74,11 +74,12 @@ namespace RendererQml
...
 		uiCard->Property("readonly property int margins", std::to_string(margin));
         uiCard->AddFunctions("signal buttonClicked(var title, var type, var data)");
-        uiCard->Property("implicitHeight", "adaptiveCardLayout.implicitHeight");
+		//1px extra height to accomodate the border of a showCard if present at the bottom
+        uiCard->Property("implicitHeight", "adaptiveCardLayout.implicitHeight+1");
 		uiCard->Property("Layout.fillWidth", "true");
 		uiCard->Property("readonly property string bgColor", context->GetRGBColor(context->GetConfig()->GetContainerStyles().defaultPalette.backgroundColor));
         uiCard->Property("color", "bgColor");
-        uiCard->Property("border.color", context->GetColor(AdaptiveCards::ForegroundColor::Default, false, false));
+		uiCard->Property("border.color", isChildCard? " bgColor" : "'#B2B2B2'");
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:82]: Is there no way to set this hardcoded color in the config?

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.h [hunk 5]
```diff
@@ -114,5 +114,12 @@ namespace RendererQml
...
+        static void CheckShowCardInActions(const std::vector<std::shared_ptr<AdaptiveCards::BaseActionElement>>& actions, std::shared_ptr<AdaptiveRenderContext> context);
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.h:120]:
```suggestion
        static void ValidateShowCardInActions(const std::vector<std::shared_ptr<AdaptiveCards::BaseActionElement>>& actions, std::shared_ptr<AdaptiveRenderContext> context);
```

File: source/qml/Library/RendererQml/AdaptiveRenderContext.cpp [hunk 6]
```diff
@@ -332,4 +332,54 @@ namespace RendererQml
...
+        {
+            return m_LastActionSetInternalIds;
+        }
+
+        return AdaptiveCards::InternalId();
+    }
+
+    void AdaptiveRenderContext::setLastActionSetInternalId(AdaptiveCards::InternalId LastActionSetInternalId)
+    {
+        m_LastActionSetInternalIds=LastActionSetInternalId;
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveRenderContext.cpp:348]:
```suggestion
        m_LastActionSetInternalIds = LastActionSetInternalId;
```

File: source/qml/Library/RendererQml/AdaptiveRenderContext.h [hunk 7]
```diff
@@ -72,7 +72,26 @@ namespace RendererQml
...
+        AdaptiveCards::InternalId getLastActionSetInternalId();
+        void setLastActionSetInternalId(AdaptiveCards::InternalId LastActionSetInternalId);
+
+        const int GetActionSetCounter();
+
+        void addToShowCardsLoaderIdsList(const std::string &loaderId);
+        const std::vector<std::string>& getShowCardsLoaderIdsList();
+
+        void setIsShowCardLastBodyElement(bool isShowCardLastBodyElement);
+        const bool getIsShowCardLastBodyElement();
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveRenderContext.h:87]:
```suggestion
        const bool isShowCardLastBodyElement();
```

File: source/qml/Library/RendererQml/AdaptiveRenderContext.h [hunk 8]
```diff
@@ -72,7 +72,26 @@ namespace RendererQml
 
         const std::string ConvertToValidId(const std::string& id);
 
+        const bool getIsShowCardInAction();
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveRenderContext.h:75]:
```suggestion
        const bool isShowCardInAction();
```

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.h [hunk 9]
```diff
@@ -44,10 +44,10 @@ namespace RendererQml
     protected:
         static void addSelectAction(const std::shared_ptr<QmlTag>& parent, const std::string& rectId, const std::shared_ptr<AdaptiveCards::BaseActionElement>& selectAction, const std::shared_ptr<AdaptiveRenderContext>& context, const bool hasBackgroundImage = false);
         static void addTextRunSelectActions(const std::shared_ptr<AdaptiveRenderContext>& context);
-        static void AddActions(std::shared_ptr<QmlTag> uiContainer, const std::vector<std::shared_ptr<AdaptiveCards::BaseActionElement>>& actions, std::shared_ptr<AdaptiveRenderContext> context);
+        static void AddActions(std::shared_ptr<QmlTag> uiContainer, const std::vector<std::shared_ptr<AdaptiveCards::BaseActionElement>>& actions, std::shared_ptr<AdaptiveRenderContext> context, bool isRemoveBottomMargin = true);
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.h:47]:
```suggestion
        static void AddActions(std::shared_ptr<QmlTag> uiContainer, const std::vector<std::shared_ptr<AdaptiveCards::BaseActionElement>>& actions, std::shared_ptr<AdaptiveRenderContext> context, bool removeBottomMargin = true);
```

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp [hunk 10]
```diff
@@ -262,10 +273,15 @@ namespace RendererQml
...
                         uiLoader->Property("id", loaderId);
-                        uiLoader->Property("x", "-margins");
+						//1px shift to avoid child card displaying over parent card's border
+                        uiLoader->Property("x", "-margins + 1");
                         uiLoader->Property("sourceComponent", componentId);
                         uiLoader->Property("visible", "false");
-						uiLoader->Property("width", Formatter() << context->getCardRootId() << ".width");
+						//2 px reduction in width to avoid child card displaying over parent card's border
+						uiLoader->Property("width", Formatter() << uiContainer->GetProperty("id") << ".width + 2*margins - 2");
+						uiLoader->Property("readonly property bool isRemoveBottomMargin", isRemoveBottomMargin ? "true" : "false");
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:282]:
```suggestion
						uiLoader->Property("readonly property bool  removeBottomMargin", removeBottomMargin ? "true" : "false");
```

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp [hunk 11]
```diff
@@ -2279,7 +2299,10 @@ namespace RendererQml
 		actionsConfig.actionAlignment = (AdaptiveCards::ActionAlignment) actionSet->GetHorizontalAlignment();
 		context->GetConfig()->SetActions(actionsConfig);
 
-		AddActions(outerContainer, actionSet->GetActions(), context);
+		auto isLastActionSet = (context->getLastActionSetInternalId() == actionSet->GetInternalId());
+		auto isShowCardInAction = context->getIsShowCardInAction();
+		auto isRemoveBottomMargin = (!isShowCardInAction && isLastActionSet);
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:2304]:
```suggestion
		auto removeBottomMargin = (!isShowCardInAction && isLastActionSet);
```

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp [hunk 12]
```diff
@@ -3196,5 +3228,80 @@ namespace RendererQml
...
+		}
+	}
+
+	const std::string AdaptiveCardQmlRenderer::RemoveBottomMarginValue(std::vector<std::string> showCardsList)
+	{
+		std::string value = "";
+
+		for (auto id : showCardsList)
+		{
+			value.append(Formatter() << "(" << id << ".visible && " << id << ".isRemoveBottomMargin) || ");
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:3293]:
```suggestion
			value.append(Formatter() << "(" << id << ".visible && " << id << ".removeBottomMargin) || ");
```

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp [hunk 13]
```diff
@@ -522,7 +539,7 @@ namespace RendererQml
 		//TODO: These color styling should come from css
         //TODO: Add hover effect
         backgroundTag->Property("color", context->GetRGBColor(context->GetConfig()->GetContainerStyles().defaultPalette.backgroundColor));
-        backgroundTag->Property("border.color", Formatter() << input->GetId() << ".activeFocus? 'black' : 'grey'");
+        backgroundTag->Property("border.color", Formatter() << input->GetId() << ".activeFocus? 'black' : '#CCCCCC'");
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:542]: Where are we getting this color code from?

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp [hunk 14]
```diff
@@ -1185,8 +1196,10 @@ namespace RendererQml
 		{
 			uiOuterRectangle->Property("radius", "3");
 		}
-		uiOuterRectangle->Property("border.color", checkbox.id + ".checked ? '#0075FF' : '767676'");
-		uiOuterRectangle->Property("color", checkbox.id + ".checked ? '#0075FF' : '#ffffff'");
+
+		auto highlightColor = context->GetColor(AdaptiveCards::ForegroundColor::Accent, false, false);
+		uiOuterRectangle->Property("border.color", Formatter() << checkbox.id << ".checked ? " << highlightColor << ": '#b0b0b0'");
+		uiOuterRectangle->Property("color", Formatter() << checkbox.id << ".checked ? " << highlightColor << " : '#ffffff'");
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:1202]: Will dark mode and light mode color be same?

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp (hunk 13 above)
Comment (rj100189) [reply source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:542]: Also looks like we are using the same color code in may places. Can we declare a const string or something at the top? Easy maintenance.

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp (hunk 3 above)
Comment (dgharde) [reply source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:125]: Thanks for pointing out. I'll make this change

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp (hunk 4 above)
Comment (dgharde) [reply source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:82]: The host config mainly has the values only for the choices of any card property (like for color its the seven choices default, dark, light, accent, etc). We may have to introduce another json or css file to store all the colors which are not adaptive card properties.

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp (hunk 13 above)
Comment (dgharde) [reply source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:542]: The color code was mentioned in [figma](https://www.figma.com/file/fi0VC0FEmETIHTTWANqMqJ/Native-B%26Cs?node-id=0%3A1). I'll make this change. Thanks

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp (hunk 14 above)
Comment (dgharde) [reply source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:1202]: This wasn't mentioned in figma. Will have to discuss this with UX

**Review by rj100189 — APPROVED** [review]

**Pull request merged**